### PR TITLE
Simplifications and Speed Improvements

### PR DESCRIPTION
**Pull request by arBmind**
This is based on #59 and adds C++14 variants where necessary.

Status:
* Tested with VS2017, VS2019 and GCC 7.3 in C++14 mode.
* Performance is still great.

Suggestions are welcome!
* [x] run CI tests in C++14 and C++17 modes
* [x] fix qmake + clang crashes (qbs + clang seems fine; I guess the settings are different)
* [x] eliminate more warnings
* [x] implement forward counting like in C++17

### CODE REVIEW

**Comment by arBmind** [comment]
It seems that qmake + clang generates crashes. I have currently have no tools to debug this further.

File: .travis.yml [hunk 1]
```diff
@@ -1,67 +1,98 @@
...
+            - libc++abi-7-dev
+            - libc++-7-dev
+            - qt512base
+      env:
+        - QT_VERSION=512
+        - CLANG_VERSION=7
 
-        # We enumerate the matrix explicitly anyway, so order it after osx.
-        - env: QT_VERSION=55 QT_SOURCE=551 GCC_VERSION=5
-        - env: QT_VERSION=56 QT_SOURCE=563 GCC_VERSION=6
```
Comment (ogoffart) [inline .travis.yml]: Maybe it is ok to drop 5.5, but I still aim at supporting 5.6, as it is a LTS

File: tests/basic/tst_basic.cpp [hunk 2]
```diff
@@ -330,7 +330,7 @@ void tst_Basic::testQNamespace()
     QCOMPARE(AnotherTU::staticMetaObject.enumerator(0).name(), "ATTestEnum");
     QCOMPARE(QMetaEnum::fromType<AnotherTU::ATTestEnum>().name(), "ATTestEnum");
 
-#ifdef __cpp_inline_variables
+#if defined (__cpp_inline_variables) && Q_CC_MSVC > 192000000
```
Comment (ogoffart) [inline tests/basic/tst_basic.cpp]: why this change?

Comment (arBmind) [reply tests/basic/tst_basic.cpp]: Simply because VS2017 cannot cope with the `inline variables`-thing you need, but defines `__cpp_inline_variables`.

**Comment by arBmind** [comment]
It seems like Gcc 5 can no longer be supported. This compiler is too eager to decay an array to just a pointer.
Everything else seems to run well. Including Qt5.5.

**Comment by ogoffart** [comment]
Great job.
This feels a lot like a complete rewrite :-)
I haven't had time to carefully the change and i would like to test them if they work fine for me. 

I guess it is OK to drop GCC 5 altough i'm curious what exactly does not work and if it is not possible to work around.

**Comment by arBmind** [comment]
It's not a complete rewrite but collection of quite some incremental improvements and refactoring.

Here is a Godbolt version of the GCC 5 issue: https://godbolt.org/z/grB7G5 
We might try to work around this by turning the `RawArray` into an `OwnArray` before assigning it to the holder. The question is, if it is worth the effort to support such an old compiler.

I hope you can review and pull it!

**Comment by ogoffart** [comment]
In the godbold, using `return Holder{ {a.strings} };` fixes it
